src/control/server/config/server.go

```diff
@@ -652,3 +651,95 @@ func (cfg *Server) validateMultiServerConfig(log logging.Logger) error {
 
 	return nil
 }
+
+var (
+	// ErrNoAffinityDetected is a sentinel error used to indicate that no affinity was detected.
+	ErrNoAffinityDetected = errors.New("no NUMA affinity detected")
+)
+
+// EngineAffinityFn defines a function which returns the NUMA node affinity of a given engine.
+type EngineAffinityFn func(logging.Logger, *engine.Config) (uint, error)
+
+func detectEngineAffinity(log logging.Logger, engineCfg *engine.Config, affSources ...EngineAffinityFn) (node uint, err error) {
+	for _, affSource := range affSources {
+		if affSource == nil {
+			return 0, errors.New("nil affinity source")
+		}
+
+		node, err = affSource(log, engineCfg)
+		if err == nil {
+			return
+		}
+
+		if err != nil && err != ErrNoAffinityDetected {
+			return
+		}
+	}
+
+	return 0, ErrNoAffinityDetected
+}
+
+func setEngineAffinity(log logging.Logger, engineCfg *engine.Config, node uint) error {
+	if engineCfg.PinnedNumaNode != nil && engineCfg.ServiceThreadCore != 0 {
```

Comment:
Does it imply that it will not be tolerated any more to have pinned_numa_node and first_core in the same engine section ?
If yes, it could be an issue as I have seen both settings used for example with the DAOS servers of the shared cluster.
However, I could probably miss something as I am not familiar with this portion of the code.

Reply:
Good to know.